protocols/system/netlink/netlink.go

```diff
@@ -0,0 +1,224 @@
+package netlink
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/bio-routing/bio-rd/config"
+	"github.com/bio-routing/bio-rd/net"
+	"github.com/bio-routing/bio-rd/route"
+	"github.com/bio-routing/bio-rd/routingtable"
+	rt "github.com/bio-routing/bio-rd/routingtable"
+	"github.com/bio-routing/bio-rd/routingtable/locRIB"
+	log "github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink"
+
+	bnet "github.com/bio-routing/bio-rd/net"
+)
+
+type NetlinkServer struct {
+	mu      sync.RWMutex
+	rib     *locRIB.LocRIB
+	options *config.Netlink
+	routes  []netlink.Route
+	routingtable.ClientManager
+}
+
+func NewNetlinkServer(options *config.Netlink, rib *locRIB.LocRIB) *NetlinkServer {
+	n := &NetlinkServer{
+		options: options,
+		rib:     rib,
+	}
+
+	n.ClientManager = routingtable.NewClientManager(n)
+
+	return n
+}
+
+func (n *NetlinkServer) Start() error {
+	log.WithField("rt_table", n.options.RoutingTable).Info("Started netlink server")
+
+	n.rib.ClientManager.Register(n)
+	go n.readKernelRoutes()
+
+	return nil
+}
+
+// Register a new client
+func (n *NetlinkServer) Register(client rt.RouteTableClient) {
+	n.ClientManager.Register(client)
+}
+
+// Register a new client with options
+func (n *NetlinkServer) RegisterWithOptions(client rt.RouteTableClient, options rt.ClientOptions) {
+	n.ClientManager.RegisterWithOptions(client, options)
+}
+
+// Unregister a client
+func (n *NetlinkServer) Unregister(client rt.RouteTableClient) {
+	n.ClientManager.Unregister(client)
+}
+
+// UpdateNewClient sends current state to a new client
+func (n *NetlinkServer) UpdateNewClient(client routingtable.RouteTableClient) error {
+	n.mu.RLock()
+	defer n.mu.RUnlock()
+
+	n.propagateChanges(make([]netlink.Route, 0), n.routes)
+
+	return nil
+}
+
+// Get the route count
+func (n *NetlinkServer) RouteCount() int64 {
+	n.mu.RLock()
+	defer n.mu.RUnlock()
+	return int64(len(n.routes))
+}
+
+// Add a path to the Kernel using netlink
+// This function is triggered by the loc_rib, cause we are subscribed as
+// client in the loc_rib
+func (n *NetlinkServer) AddPath(pfx net.Prefix, path *route.Path) error {
+	log.WithFields(log.Fields{
+		"Prefix": pfx.String(),
+	}).Debug("AddPath to netlink")
+
+	err := netlink.RouteAdd(n.createRoute(pfx, path))
```

Comment:
Maybe use https://godoc.org/github.com/vishvananda/netlink#RouteReplace instead of RouteAdd? Replace will do an add if not present, else an replacement

Reply:
Awesome! Thanks

Reply:
I wouldn't do that. Without checking that the route is from use "proto bio", we might overwrite/replace a route that we did not put into the kernel table, which I would consider a hard bug. So IMO there are some more checks required which is a second strong argument for another bio-internal table which represents the kernel table.

Reply:
This is true. Could we just reuse https://github.com/bio-routing/bio-rd/blob/master/routingtable/table.go#L12 ? We use this table already in multiple components like adj_rib_out: https://github.com/bio-routing/bio-rd/blob/master/routingtable/adjRIBOut/adj_rib_out.go